File: src/use_cases/evaluate.py
from datetime import datetime
from typing import Dict, Any

from src.infrastructure.ai import Curator
from src.infrastructure.repositories import FileRepository, SupabaseRepository
from src.infrastructure.settings import settings


class EvaluateDailyContentUseCase:
    def __init__(
        self,
        curator: Curator | None = None,
        file_repository: FileRepository | None = None,
        storage: SupabaseRepository | None = None,
    ):
        self._curator = curator or Curator()
        self._files = file_repository or FileRepository()
        self._storage = storage or SupabaseRepository()

    def execute(self, date_str: str = None) -> Dict[str, Any] | None:
        target_date = date_str or datetime.now().strftime("%Y%m%d")
        
        summary_path = settings.summary_dir / f"{target_date}_summary.txt"
        novel_path = settings.novel_out_dir / f"{target_date}.md"

        if not summary_path.exists():
            print(f"Summary not found for {target_date}")
            return None
            
        if not novel_path.exists():
            print(f"Novel not found for {target_date}")
            return None

        summary_text = self._files.read(str(summary_path))
        novel_text = self._files.read(str(novel_path))

        print(f"Evaluating content for {target_date}...")
        result = self._curator.evaluate(summary_text, novel_text)
        
        print(f"  Faithfulness: {result.get('faithfulness_score')}/5")
        print(f"  Quality: {result.get('quality_score')}/5")
        
        self._files.save_evaluation(result, target_date)
        self._storage.sync()
        
        return result

File: src/use_cases/evaluate.rs
use crate::domain::Curator;
use crate::infrastructure::api::SupabaseClient;
use std::fs;
use std::path::Path;
use tracing::info;

pub struct EvaluateDailyContentUseCase {
    curator: Box<dyn Curator>,
    supabase: Option<SupabaseClient>,
}

impl EvaluateDailyContentUseCase {
    pub fn new(curator: Box<dyn Curator>, supabase: Option<SupabaseClient>) -> Self {
        Self { curator, supabase }
    }

    pub async fn execute(&self, date: &str) {
        let summary_path = crate::domain::constants::SUMMARY_FILE_TEMPLATE.replace("{}", date);
        let novel_path = crate::domain::constants::NOVEL_FILE_TEMPLATE.replace("{}", date);

        if !Path::new(&summary_path).exists() || !Path::new(&novel_path).exists() {
            panic!("Summary or Novel not found for {}", date);
        }

        let summary_text = fs::read_to_string(summary_path).unwrap();
        let novel_text = fs::read_to_string(novel_path).unwrap();

        info!("Evaluating content for {}...", date);
        let result = self.curator.evaluate(&summary_text, &novel_text).await;

        let eval_path = crate::domain::constants::EVALUATION_FILE_TEMPLATE.replace("{}", date);
        if let Some(parent) = Path::new(&eval_path).parent() {
            fs::create_dir_all(parent).unwrap();
        }
        fs::write(
            &eval_path,
            serde_json::to_string_pretty(&result).unwrap(),
        )
        .unwrap();
        info!("Evaluation saved to {}", eval_path);

        if let Some(ref supabase) = self.supabase {
            info!("Syncing evaluation to Supabase...");
            let data = serde_json::json!({
                "date": date,
                "target_type": "novel",
                "score": result.quality_score,
                "reasoning": result.reasoning
            });
            supabase.upsert("evaluations", &data).await.unwrap();
        }

        }
}
